lib/elixir/lib/map.ex

```diff
@@ -211,13 +220,16 @@ defmodule Map do
   """
   @spec take(map, [key]) :: map
   def take(map, keys) do
-    Enum.reduce(keys, [], fn key, acc ->
-      case fetch(map, key) do
-        {:ok, value} -> [{key, value} | acc]
-        :error -> acc
-      end
-    end)
-    |> :maps.from_list
+    do_take(keys, map, [])
+  end
```

Comment:
For some reason this approach produces a much more pronounced effect with `Map.take`

```
Map.take list manual recursion         1000   1057.48 µs/op
Map.take range manual recursion        1000   1583.72 µs/op
Map.take list master                   1000   2135.62 µs/op
Map.take range master                  1000   2571.30 µs/op
```

Code: 
```
  def take(map, keys) when is_list(keys) do
    do_take(keys, map, [])
  end
  def take(map, keys) do
    keys
    |> Enum.reverse
    |> do_take(map, [])
  end

  defp do_take([], _map, acc), do: :maps.from_list(acc)
  defp do_take([key | rest], map, acc) do
    acc = case fetch(map, key) do
      {:ok, value} -> [{key, value} | acc]
      :error -> acc
    end
    do_take(rest, map, acc)
  end
```